vperfetto-min.cpp

```diff
@@ -0,0 +1,148 @@
+
+static void initPerfetto() {
+    if (!sPerfettoInitialized) {
+        ::perfetto::TracingInitArgs args;
+        args.backends |= ::perfetto::kInProcessBackend;
+        ::perfetto::Tracing::Initialize(args);
+        ::perfetto::TrackEvent::Register();
+        sPerfettoInitialized = true;
+    }
+}
+
+bool useFilenameByEnv(const char* s) {
+    return s && ("" != std::string(s));
+}
+
+VPERFETTO_EXPORT void vperfetto_min_startTracing(const char* filename) {
+    sTraceConfig.hostFilename = filename;
+
+    // Ensure perfetto is actually initialized.
+    initPerfetto();
+
+    if (!sTracingSession) {
+        fprintf(stderr, "%s: Tracing begins================================================================================\n", __func__);
+        fprintf(stderr, "%s: Configuration:\n", __func__);
+        fprintf(stderr, "%s: host filename: %s (possibly set via $VPERFETTO_HOST_FILE)\n", __func__, sTraceConfig.hostFilename);
+
+        auto desc = ::perfetto::ProcessTrack::Current().Serialize();
+        desc.mutable_process()->set_process_name("VirtualMachineMonitorProcess");
+        ::perfetto::TrackEvent::SetTrackDescriptor(::perfetto::ProcessTrack::Current(), desc);
+
+        ::perfetto::TraceConfig cfg;
+        ::perfetto::protos::gen::TrackEventConfig track_event_cfg;
+        // TODO: Should this be another parameter?
+        cfg.add_buffers()->set_size_kb(1024 * 100);  // Record up to 100 MiB.
+        auto* ds_cfg = cfg.add_data_sources()->mutable_config();
+        ds_cfg->set_name("track_event");
+        ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+
+        sTracingSession = ::perfetto::Tracing::NewTrace();
+        sTracingSession->Setup(cfg);
+        sTracingSession->StartBlocking();
+        sTraceConfig.tracingDisabled = false;
+    }
+}
+
+VPERFETTO_EXPORT void vperfetto_min_endTracing() {
+    if (sTracingSession) {
+        sTraceConfig.tracingDisabled = true;
+
+        // Don't disable again if we are saving.
+        if (sTraceConfig.saving) return;
+        sTraceConfig.saving = true;
+
+        sTracingSession->StopBlocking();
+        sTraceProgress.hostTrace = sTracingSession->ReadTraceBlocking();
+
+        fprintf(stderr, "%s: Tracing ended================================================================================\n", __func__);
+        fprintf(stderr, "%s: Saving trace to disk. Configuration:\n", __func__);
+        fprintf(stderr, "%s: host filename: %s\n", __func__, sTraceConfig.hostFilename);
+
+        sTracingSession.reset();
+
+       const char* hostFilename = sTraceConfig.hostFilename;
+       {
+           std::ofstream hostFile(sTraceConfig.hostFilename, std::ios::out | std::ios::binary);
+           hostFile.write(sTraceProgress.hostTrace.data(), sTraceProgress.hostTrace.size());
+       }
+       sTraceConfig.saving = false;
+    }
+}
+
+VPERFETTO_EXPORT void vperfetto_min_beginTrackEvent(const char* eventName) {
```

Comment:
I think I will need something like this, to support multiple tracks:

https://gitlab.freedesktop.org/tomeu/mesa/-/blob/perfetto-tracing/src/util/u_tracing.cpp#L96

I think we probably can find a set of common tracks that work for everybody: OpenGL, Vulkan, Driver, VMM, anything else? Gallium events can be probably put into Driver. Virglrenderer events can be probably put into the VMM track.

I went with one category per track, but I think a single "gfx" category will be fine.

Reply:
Cool. Added some graphics/VMM categories in 51faaf63bca960f12bb7dd02eb75f19b82421353 and allowed track events in each one in 3a94e9eb55bddbad0a6248a9af300018f7892bd0.